gen-bpf.py

```diff
@@ -0,0 +1,300 @@
+    void *buf = NULL;
+    struct event_t event = { .id = %d };
+
+""" % (build_tracer_name(metadata), metadata['id'])
+
+    block_field_set = block_fields.get(metadata["fully_specified_probe_name"], set())
+
+    i = 0
+    args = metadata['args']
+    str_i = 0
+    int_i = 0
+    while i < len(args):
+        arg = args[i]
+        arg_name = arg['name']
+        arg_type = arg['type']
+        c += "    // %s %s\n" % (arg_type, arg_name)
+        if arg_name in block_field_set:
+            c += "    // (ignored because it's in the block list)\n"
+            i += 1
+            continue
+
+        if is_str_type(arg_type):
+            c += "    bpf_usdt_readarg(%d, ctx, &buf);\n" % (i+1)
+            # Use `sizeof(buf)` instead of a length variable, because older kernels
+            # do not accept a variable for `bpf_probe_read()`'s length parameter.
+            c += "    bpf_probe_read(&event.s%d, sizeof(event.s%d), buf);\n" % (str_i, str_i)
+            str_i += 1
+        elif is_ptr_type(arg_type):
+            c += "    %s %s = {};\n" % (arg_type.replace("*", ""), arg_name)
+            c += "    bpf_usdt_readarg(%d, ctx, &buf);\n" % (i+1)
+            c += "    bpf_probe_read(&%s, sizeof(%s), buf);\n" % (arg_name, arg_name)
+            st_name = strip_typename(arg_type)
+            for st_key, st_valtype in st_map[st_name].items():
+                c += "    event.i%d = %s.%s; /* %s */\n" % (int_i, arg_name, st_key, st_valtype)
+                int_i += 1
+        else:
+            c += "    bpf_usdt_readarg(%d, ctx, &event.i%d);\n" % (i+1, int_i)
+            int_i += 1
+        i += 1
+    diff = ""
+    if str_i > 0:
+        diff = " - %d" % (str_i * 32)
+    c += """
+    if (events.perf_submit(ctx, &event, sizeof(event)%s) != 0)
+        bpf_trace_printk("failed to perf_submit\\n");
+
+    return 0;
+}\n""" % diff
+    return c
+
+
+st_map = {}
+for (st_name, content) in re.findall(r'struct\s+([a-zA-Z0-9_]+)\s*\{([^}]*)\}', struct_decl, flags = re_flags):
+    st = st_map[st_name] = {}
+    for (ctype, name, is_array) in re.findall(r'(\w+[^;]*[\w\*])\s+([a-zA-Z0-9_]+)(\[\d+\])?;', content, flags = re_flags):
+        if "dummy" in name:
+            continue
+        st[name] = ctype + is_array
+
+probe_metadata = OrderedDict()
+probe_id2metadata = OrderedDict()
+max_ints = 0
+max_strs = 0
+
+id = 0
+for (name, args) in re.findall(r'\bprobe\s+([a-zA-Z0-9_]+)\(([^\)]+)\);', matched.group('probes'), flags = re_flags):
+    arg_list = re.split(r'\s*,\s*', args, flags = re_flags)
+    id += 1
+    metadata = {
+        "id": id,
+        "provider": provider,
+        "name": name,
+        "fully_specified_probe_name": "%s:%s" % (provider, name),
+    }
+    probe_id2metadata[id] = metadata
+    probe_metadata[name] = metadata
+    args = metadata['args'] = map(
+        lambda arg: re.match(r'(?P<type>\w[^;]*[^;\s])\s*\b(?P<name>[a-zA-Z0-9_]+)', arg, flags = re_flags).groupdict(),
+        arg_list)
```

Comment:
`map` built-in function returns `map` type since Python3 while Python2 `map` returns `list` type. And `map` type cannot be counted length by `len`. `gen-bpf.py` does not work with python3

```
Traceback (most recent call last):
  File "./gen-bpf.py", line 207, in <module>
    bpf += build_tracer(metadata)
  File "./gen-bpf.py", line 84, in build_tracer
    while i < len(args):
TypeError: object of type 'map' has no len()
```

This `map` should be wrapped with `list` or replace it with list comprehension.

```
% python2
Python 2.7.17 (default, Nov  7 2019, 10:07:09)
[GCC 9.2.1 20191008] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> a = map(lambda a: a, range(1))
>>> a.__class__
<type 'list'>
% python3
Python 3.7.5 (default, Nov 20 2019, 09:21:52)
[GCC 9.2.1 20191008] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> a = map(lambda a: a, range(1))
>>> a.__class__
<class 'map'>
>>> len(a)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: object of type 'map' has no len()
```

warp with `list`

```diff
diff --git a/gen-bpf.py b/gen-bpf.py
index 20e92eb..649d6f2 100755
--- a/gen-bpf.py
+++ b/gen-bpf.py
@@ -146,9 +146,9 @@ for (name, args) in re.findall(r'\bprobe\s+([a-zA-Z0-9_]+)\(([^\)]+)\);', matche
     }
     probe_id2metadata[id] = metadata
     probe_metadata[name] = metadata
-    args = metadata['args'] = map(
+    args = metadata['args'] = list(map(
         lambda arg: re.match(r'(?P<type>\w[^;]*[^;\s])\s*\b(?P<name>[a-zA-Z0-9_]+)', arg, flags = re_flags).groupdict(),
-        arg_list)
+        arg_list))

     # args map is a flat arg list
     args_map = metadata['args_map'] = OrderedDict()
```

Use list comprehension

```diff
diff --git a/gen-bpf.py b/gen-bpf.py
index 20e92eb..7a41406 100755
--- a/gen-bpf.py
+++ b/gen-bpf.py
@@ -146,9 +146,7 @@ for (name, args) in re.findall(r'\bprobe\s+([a-zA-Z0-9_]+)\(([^\)]+)\);', matche
     }
     probe_id2metadata[id] = metadata
     probe_metadata[name] = metadata
-    args = metadata['args'] = map(
-        lambda arg: re.match(r'(?P<type>\w[^;]*[^;\s])\s*\b(?P<name>[a-zA-Z0-9_]+)', arg, flags = re_flags).groupdict(),
-        arg_list)
+    args = metadata['args'] = [re.match(r'(?P<type>\w[^;]*[^;\s])\s*\b(?P<name>[a-zA-Z0-9_]+)', arg, flags = re_flags).groupdict() for arg in arg_list]

     # args map is a flat arg list
     args_map = metadata['args_map'] = OrderedDict()
```

Reply:
Thanks. As of `master` branch, it need working with Python 2.7, but now it can discard 2.7 compatibility, so I'll just make it work with Python 3.5+ (FYI Python 3.5 is the default on Ubuntu 16).

Reply:
done in 64eca28b9bd7ac56ccf73c1394eea8f9b7b93c7c

Reply:
Great stuff! this program looks entirely dedicated to string processing, so it could be written in anything well supported, say... Perl right? (I'm joking, this is not a suggestion).

Reply:
Indeed, I would use Perl if I wrote it from scratch!

Reply:
Y'know, it makes sense if we think about eventually moving to H2O. All the auxiliary scripts are Perl over there. Food for thought!